modules/javafx.web/src/test/java/test/javafx/scene/web/MathMLRenderTest

```diff
@@ -0,0 +1,85 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.javafx.scene.web;
+
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+
+public class MathMLRenderTest extends TestBase {
+
+    @Test public void testTokenHeight() throws Exception {
+
+        // Document test
+        String mathmlContent
+            = "<math display=\"block\">"
+            + "   <mrow>"
+            + "      <mi>x</mi>"
+            + "      <mo>=</mo>"
+            + "      <mfrac>"
+            + "         <mrow>"
+            + "            <mo>−</mo>"
+            + "            <mi>b</mi>"
+            + "            <mo>±</mo>"
+            + "            <msqrt>"
+            + "               <mrow>"
+            + "                  <msup>"
+            + "                     <mi>b</mi>"
+            + "                     <mn>2</mn>"
+            + "                  </msup>"
+            + "                  <mo>−</mo>"
+            + "                  <mn>4</mn>"
+            + "                  <mi>a</mi>"
+            + "                  <mi>c</mi>"
+            + "               </mrow>"
+            + "            </msqrt>"
+            + "         </mrow>"
+            + "         <mrow>"
+            + "            <mn>2</mn>"
+            + "            <mi>a</mi>"
+            + "         </mrow>"
+            + "      </mfrac>"
+            + "   </mrow>"
+            + "</math>";
+
+        String htmlBody = "<!doctype html>"
+            + "<html>"
+            + "   <body>"
+            + "      <p>"
+            + mathmlContent
+            + "      </p>"
+            + "   </body>"
+            + "</html>";
+            
+        loadContent(htmlBody);
+
+        int height = (int) executeScript(
+            "elements = document.getElementsByTagName('mo');"
```

Comment:
So you are testing the height of the first `mo` tag which seems to be `=`.

Why can't use a very basic mathml markup which might be enough to test the problematic scenario? Something like below?

```
<body>
<math id='math'>
      <mfrac>
        <mn>1</mn>
        <mn>2</mn>
      </mfrac>
</math>
</body>
```

then,
```
final int EXPECTED_MINIUM_HEIGHT = 25;
int height = executeScript ("document.getElementById('html').clientHeight");
assertTrue("MathML token height is " + height, height > EXPECTED_MINIUM_HEIGHT);
```

Reply:
You could change but it will be regret that :-( :
-I used a basic element because the problem occured for any the element, text token element include.
So the patch simply solves a character height bug, but it's difficult to see it only with a text element. In a text element the only sign was the cursor height.
- Thus we don't need a formula to test it. I used this opportunity to show how to do to test any others token elements.
- And the quadratic formula is a reference to the first bug report in 2015.
- But, it's also because, if you test with complexe formula the height could be over 25 : when formula contains a fraction the box size is not equal to 0 you must take in account the fraction line tickness. So if you've got fraction in a fraction ... the size grows up.

As I esplained in #71 the OpenJFX and WebCore are clean. I never had doubt about this.
One who wrote the code thought logically that the rendering of a simple MathML text like "Hello" is made exactly with the same logic used for same text in a html paragraph. Why it should be different ?

In fact WebCore and most of all others mathematic editor use a rendering way I hate. In their logic they can't used ascents and descents because of stretchy assembled glyphs like ( [ { ... : for example to construct a big { you need to assemble 3 characters (like bricks for a wall).
If you use the glyph ascent or descent , you'll see gap between characters, because it is the common ascent or descent for all glyph and not the ascent or descent of the glyph you're using.

I understood where was the problem when I read his comment "//That is OK! platformWidthForGlyph impl is enough." He tought that getAscent() and getDescent() was enough.

I aware of this problem because I'm building my own math editor. I prefer draw these stretchy glyphs by myself. I sure that provide a better experience to users. The Swing API allows this but to use it, I'had been send several bug reports and build a patch. I sent the patch in 2012 and it has been accepted in 2017 when Sergey Bylokhov accepted to sponsor me.

You said that present WebKit version is 2.18. But I saw that some code are older. What is your project for the next WebKit version. There are many changes in MathML support, they improved RTL support and other things. Because of this PR work I haven't time to look further inside.

Reply:
Since it is a headless test, I prefer to have the test case as simple as possible. Basically the test case should able to catch the future regressions. Even though you render the quadratic formula, you end up checking only the height of '='. Then why can't just have a simple element with only '=' (or anything which you wish). 

(This is my only concern right now, I request @kevinrushforth to take a look)

Reply:
All that work just for "=" :-D
As I said the quadric formula is a reference to the bug report Kevin treated in 2015 :-)
Of course we can.
Your are right a headless test doesn't need more.
I let you decide.

![quadratic_formula](https://user-images.githubusercontent.com/19194678/41858334-e6613eaa-7899-11e8-887e-2cbd3e2a297d.png)

Reply:
According your request the test could be :
But I'm not at home and I can't test it before Friday.

```
/*
 * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package test.javafx.scene.web;

import static org.junit.Assert.assertTrue;
import org.junit.Test;

public class MathMLRenderTest extends TestBase {

    @Test public void testTokenHeight() throws Exception {

        // Document test
        String htmlBody = "<!doctype html>"
            + "<html>"
            + "   <body>"
            + "      <p>"
            +  "        <math display=\"block\">"
            + "            <mrow>"
            + "               <mo>=</mo>"
            + "            </mrow>"
            + "         </math>";
            + "      </p>"
            + "   </body>"
            + "</html>";
            
        loadContent(htmlBody);

        int height = (int) executeScript(
            "elements = document.getElementsByTagName('mo');"
            + "element = elements[0].clientHeight;"
        );

        assertTrue("MathML token height must be greater than " + height, height > 1);      
    }
}
```